### PR TITLE
Fix CPIO and TAR padding

### LINKED ISSUE #58: Integration tests trigger unknown chunks warning messages.
Our integration tests trigger warnings for unknown chunks when parsing CPIO and TAR formats. This should not happen.

It's probably an end offset miscalculation due to padding, or test files not respecting the standard.

### PR DESCRIPTION
**Pull request by QKaiser**
We fixed padding errors in both CPIO and TAR handlers that were identified during pytest runs due to unknown chunks of null bytes being identified out of nowhere.

Fix #58 

This was due to padding issues described below:

### CPIO

> (Copy Out.) Read the standard input to obtain a list of pathnames and copy those files onto the standard output together with pathname and status information. Output is padded to a 512-byte boundary.

We're now taking the 512 bytes padding into consideration.

### TAR

> The data in an archive is grouped into blocks, which are 512 bytes. Blocks are read and written in whole number multiples called records. The number of blocks in a record (i.e., the size of a record in units of 512 bytes) is called the blocking factor.

So the overall size of a TAR archive is padded to the block factor, which is set to 20 by default. The block factor padding is therefore 512 * 20 = 10240.

We're now padding the whole chunk to the block factor padding size.

Each format handler and integration tests fix has been put into its own dedicated commit (CPIO, TAR).

### CODE REVIEW

**Comment by kukovecz** [comment]
For a simply test, I believe this should be `0`:

```shell
$ pytest -s tests/test_handlers.py | grep "Found unknown Chunks" | wc -l
62
```

At least, the concern in #58 that our tests are finding unknown chunks.

**Comment by QKaiser** [comment]
> For a simply test, I believe this should be `0`:
> 
> ```shell
> $ pytest -s tests/test_handlers.py | grep "Found unknown Chunks" | wc -l
> 62
> ```
> 
> At least, the concern in #58 that our tests are finding unknown chunks.

These are coming from the text files inside the archives we're currently analyzing.

**Comment by kissgyorgy** [comment]
> As far as I can see the same thing was done for both `tar` and `cpio`. Something like:
> ...
> Should not we have some common function for this in `Handler` (or `StructHandler` ?), like:
> ...
> So this could be called from the handlers more easily? Would this make sense?

I agree that this should be a helper function, but I would rather put it into file_utils.py as a simple function instead.

**Comment by kissgyorgy** [comment]
Also I think that calculation is exactly the same as `round_up`, maybe we should rename that function to `pad_bytes` or something?

**Comment by QKaiser** [comment]
> Also I think that calculation is exactly the same as `round_up`, maybe we should rename that function to `pad_bytes`, `add_padding` or something?

They look similar but they don't perform the same calculation. `round_up` calculates the upper size aligned on padding boundary, while `pad_bytes` calculates the amount of bytes required to pad a specific offset to reach a specific alignment.

With an offset of 502 and alignment of 512, `round_up` would return 512, but `pad_bytes` would return 10 :)

I'll create that function in file_utils and write some tests.

**Comment by QKaiser** [comment]
@kissgyorgy @kukovecz switched to a dedicated function in ´file_utils` called `pad_bytes` with supporting test cases, both CPIO and TAR handler use that function now.

**Review by kissgyorgy — CHANGES_REQUESTED** [review]

File: unblob/handlers/archive/tar.py [hunk 1]
```diff
@@ -23,6 +23,11 @@
 END_BLOCK_SIZE = BLOCK_SIZE * 2
 END_BLOCK = b"\x00" * END_BLOCK_SIZE
 
+# Default block factor for tar on Unix systems and default for Python's
+# tarfile (see https://docs.python.org/3/library/tarfile.html#tarfile.open).
+BLOCK_FACTOR = 20
```
[the remark this answers is not in the document]
Comment (QKaiser) [reply unblob/handlers/archive/tar.py]: Good catch, moved to using `tarfile.RECORDSIZE`.

File: unblob/handlers/archive/cpio.py [hunk 2]
```diff
@@ -56,6 +57,13 @@ def calculate_chunk(
             # need to round up again.
             offset += self._pad_content(header, c_filesize, c_namesize)
 
+        # Add padding that could exists between the cpio trailer and the end-of-file.
+        # cpio aligns the file to 512 bytes
+        padding = (
+            self._FILE_PAD_ALIGN - (offset % self._FILE_PAD_ALIGN)
+        ) % self._FILE_PAD_ALIGN
+        offset += padding
+
```
[the remark this answers is not in the document]
Comment (QKaiser) [reply unblob/handlers/archive/cpio.py:63]: No. See https://github.com/IoT-Inspector/unblob/pull/64#issuecomment-982461311.

File: unblob/handlers/archive/tar.py [hunk 3]
```diff
@@ -85,8 +90,9 @@ def calculate_chunk(
 
         file.seek(start_offset)
         end_offset = _get_tar_end_offset(file)
+        padding = (BLOCK_PADDING - (end_offset % BLOCK_PADDING)) % BLOCK_PADDING
```
[the remark this answers is not in the document]
Comment (QKaiser) [reply unblob/handlers/archive/tar.py]: Moved to _get_tar_end_offset.

**Comment by QKaiser** [comment]
@kissgyorgy moved to `round_up` instead of `pad_bytes` since our discussion on the subject :)

File: unblob/handlers/archive/tar.py [hunk 4]
```diff
@@ -31,6 +31,7 @@ def _get_tar_end_offset(file: io.BufferedIOBase):
     last_member = tf.getmembers()[-1]
     last_file_size = BLOCK_SIZE * (1 + (last_member.size // BLOCK_SIZE))
     end_offset = last_member.offset + HEADER_SIZE + last_file_size + END_BLOCK_SIZE
```
Comment (kissgyorgy) [inline unblob/handlers/archive/tar.py:33]: `last_offset` would be a better name, or name the padded offset `padded_end_offset` or `end_offset_padded` instead.

Comment (QKaiser) [reply unblob/handlers/archive/tar.py]: Fixed.

**Review by kissgyorgy — APPROVED** [review]

**Pull request merged**